### PR TITLE
changes in the hadronBlock 

### PR DESCRIPTION
**Pull request by MRD2F**

### CODE REVIEW

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 1]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
 
     tensorflow::Tensor createHadronsBlockInputs(const TauType& tau, const reco::Vertex& pv, double rho,
                                                 const pat::PackedCandidateCollection& pfCands,
-                                                const CellGrid& grid, bool is_inner)
+                                                CellGrid& grid, bool is_inner)
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: should be `const CellGrid& grid`

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 2]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        using namespace HadronBlockInputs;
+
+        int eta_index, phi_index;
+        const int max_eta_index = grid.maxEtaIndex(), max_phi_index = grid.maxPhiIndex();
+        for(eta_index = -max_eta_index; eta_index <= max_eta_index; ++eta_index) {
+            for(phi_index = -max_phi_index; phi_index <= max_phi_index; ++phi_index) {
+                    eta_index =+  max_eta_index;
+                    phi_index =+ max_phi_index;
+            }
+        }
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: these loops are not needed

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 3]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+
+        tensorflow::Tensor inputs(tensorflow::DT_FLOAT, {1, eta_index, phi_index, NumberOfInputs});
+        inputs.flat<float>().setZero();
 
+        const CellIndex cellIndex{eta_index, phi_index};
+        Cell& cell = grid[cellIndex];
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: how can you have cellIndex outside of the for? it is wrong

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 4]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        auto index_chH = cell.at(CellObjectType::PfCand_chargedHadron);
+        auto index_nH = cell.at(CellObjectType::PfCand_neutralHadron);
+        auto n_index_chH = cell.size();
+
+        for(const auto& cell : grid) {
+            const auto& get = [&](int var_index) -> float& {
+                return inputs.tensor<float,4>()(0,eta_index,phi_index,var_index);
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc:713]: why eta_index and phi_index don't depend on cell?

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 5]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+            };
+
+        get(rho) = GetValueNorm(rho, 21.49f, 9.713f);
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: if there is no charged and no neutral pfCand in the cell, also the global vars shouldn't be filled.

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 6]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        get(tau_pt) =  GetValueLinear(tau.polarP4().pt(), 20.f, 1000.f, true);
+        get(tau_eta) = GetValueLinear(tau.polarP4().eta(), -2.3f, 2.3f, false);
+        get(tau_inside_ecal_crack) = GetValue(isInEcalCrack(tau.p4().eta()));
+
+        const bool valid = n_index_chH != 0;
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: this is a wrong condition to check the validity

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 7]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        get(pfCand_chHad_valid) = valid;
+        get(pfCand_chHad_rel_pt) = valid ? GetValueNorm(pfCands.at(index_chH).polarP4().pt() / tau.p4().pt(),
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: to improve the speed, do all set inside `if(valid)`, instead using `?` operator in each line. You already set to zero the default values.

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 8]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        const int max_eta_index = grid.maxEtaIndex(), max_phi_index = grid.maxPhiIndex();
+        for(eta_index = -max_eta_index; eta_index <= max_eta_index; ++eta_index) {
+            for(phi_index = -max_phi_index; phi_index <= max_phi_index; ++phi_index) {
+                    eta_index =+  max_eta_index;
+                    phi_index =+ max_phi_index;
+            }
+        }
+
+        tensorflow::Tensor inputs(tensorflow::DT_FLOAT, {1, eta_index, phi_index, NumberOfInputs});
+        inputs.flat<float>().setZero();
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc:706]: does it raise some warnings? in the past, similar call resulted in some warning during the compilation.

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 9]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        get(pfCand_chHad_vertex_dx_tauFL) = valid ? GetValueNorm(pfCands.at(index_chH).vertex().x() - pv.position().x() -
+            tau.flightLength().x(), -0.0014f, 1.93f) : 0;
+        get(pfCand_chHad_vertex_dy_tauFL) = valid ? GetValueNorm(pfCands.at(index_chH).vertex().y() - pv.position().y() -
+            tau.flightLength().y(), 0.0022f, 1.948f) : 0;
+        get(pfCand_chHad_vertex_dz_tauFL) = valid ? GetValueNorm(pfCands.at(index_chH).vertex().z() - pv.position().z() -
+            tau.flightLength().z(), -0.0138f, 8.622f) : 0;
+
+        const bool hasTrackDetails = valid && pfCands.at(index_chH).hasTrackDetails() == 1;
+        get(pfCand_chHad_hasTrackDetails) = hasTrackDetails;
+        get(pfCand_chHad_dxy) = hasTrackDetails ? GetValueNorm(std::abs(pfCands.at(index_chH).dxy()) /
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: same here: call get inside the `if(hasTrackDetails)`

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 10]
```diff
@@ -667,11 +697,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        get(pfCand_chHad_hcalFraction) = valid ? GetValue(pfCands.at(index_chH).hcalFraction()) : 0;
+        get(pfCand_chHad_rawCaloFraction) = valid ?
+            GetValueLinear(pfCands.at(index_chH).rawCaloFraction(), 0.f, 2.6f, true) : 0;
+
+        get(pfCand_nHad_valid) = valid;
+        get(pfCand_nHad_rel_pt) = valid ? GetValueNorm(pfCands.at(index_nH).polarP4().pt() / tau.polarP4().pt(),
+            is_inner ? 0.3163f : 0.0502f, is_inner ? 0.2769f : 0.4266f) : 0;
+        get(pfCand_nHad_deta) = valid ? GetValueLinear(pfCands.at(index_nH).polarP4().eta() - tau.polarP4().eta(),
+            is_inner ? -0.1f : -0.5f, is_inner ? 0.1f : 0.5f, false) : 0;
+        get(pfCand_nHad_dphi) = valid ? GetValueLinear(ROOT::Math::VectorUtil::DeltaPhi(pfCands.at(index_nH).polarP4(), tau.polarP4()),
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: order of arguments in DeltaPhi looks wrong, it should be consistent with how it is done in the training code.

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 10 above)
Comment (MRD2F) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: should be done with the function DeltaPhi defined in https://github.com/kandrosov/TauML/blob/1738f430a4d9ffb28fa06566c6fbd974e596659e/Analysis/source/TrainingTupleProducer.cxx#L161?

Comment (kandrosov) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: you can use `dPhi` method defined in https://github.com/MRD2F/cmssw/blob/92bd59f045e15b1623ae153fade59a528c522f82/RecoTauTag/RecoTau/plugins/DeepTauId.cc#L144 (but need to change the order of arguments inside)

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 6 above)
Comment (MRD2F) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: first I think the condition was defined here:  https://github.com/kandrosov/TauML/blob/1738f430a4d9ffb28fa06566c6fbd974e596659e/Analysis/source/TrainingTupleProducer.cxx#L366. :

const auto& index_set = cell[type];
n_total = index_set.size();

In this case, is the size of what? the size of cell[type] it's already and int

Comment (kandrosov) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: now in each cell there is only one candidate, so, should check for the presence of a candidate with the given type.

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 8 above)
Comment (MRD2F) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc:706]: No, This part doesn't give any warnings for now

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 11]
```diff
@@ -84,15 +84,51 @@ namespace TauBlockInputs {
...
 }
 
 namespace MuonBlockInputs {
-    enum vars {rho, tau_pt, tau_eta, tau_inside_ecal_crack};
+    enum vars {rho = 0, tau_pt, tau_eta, tau_inside_ecal_crack};
 }
 
 namespace HadronBlockInputs {
-    enum vars {rho, tau_pt, tau_eta, tau_inside_ecal_crack};
+    enum vars {rho = 0, tau_pt, tau_eta, tau_inside_ecal_crack, NumberOfInputs, pfCand_chHad_valid,
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: `NumberOfInputs` should be the last enum element

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 12]
```diff
@@ -669,9 +699,108 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
     {
-        return tensorflow::Tensor();
-    }
+        using namespace dnn_inputs_2017_v2;
+        using namespace HadronBlockInputs;
+
+        int eta_index, phi_index;
+        const int max_eta_index = grid.maxEtaIndex(), max_phi_index = grid.maxPhiIndex();
+
+        tensorflow::Tensor inputs(tensorflow::DT_FLOAT, {1, eta_index, phi_index, NumberOfInputs});
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: should use `grid.nCellsEta` and `grid.nCellsPhi`

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 13]
```diff
@@ -669,9 +699,108 @@ class DeepTauId : public deep_tau::DeepTauBase {
                                                 const pat::PackedCandidateCollection& pfCands,
                                                 const CellGrid& grid, bool is_inner)
     {
-        return tensorflow::Tensor();
-    }
+        using namespace dnn_inputs_2017_v2;
+        using namespace HadronBlockInputs;
+
+        int eta_index, phi_index;
+        const int max_eta_index = grid.maxEtaIndex(), max_phi_index = grid.maxPhiIndex();
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: these two lines must be removed.

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 14]
```diff
@@ -669,9 +699,108 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+        inputs.flat<float>().setZero();
+
+        for(const auto& cell : grid) {
+            eta_index = cell.first.eta;
+            phi_index = cell.first.phi;
 
+            eta_index =+  max_eta_index;
+            phi_index =+ max_phi_index;
+
+            const CellIndex cellIndex{eta_index, phi_index};
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: - why do you need `cellIndex` object?
- to get eta,phi index in tensor better to call `grid. getEtaTensorIndex(cell.first)` (similar for phi)

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 15]
```diff
@@ -669,9 +699,108 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+
+            const auto& get = [&](int var_index) -> float& {
+                return inputs.tensor<float,4>()(0,eta_index,phi_index,var_index);
+            };
+
+            auto cell_map = std::get<1>(cell);
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: - better cell.second
- should be const auto&

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 16]
```diff
@@ -669,9 +699,108 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+            size_t index_chH = cell_map[CellObjectType::PfCand_chargedHadron];
+            size_t index_nH = cell_map[CellObjectType::PfCand_neutralHadron];
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: should check presence first, can't index const map with `[]`

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 17]
```diff
@@ -669,9 +699,108 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+            if(index_chH != 0 && index_nH != 0){
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: wrong condition: index = 0 corresponds to a valid cell position. Same below

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 18]
```diff
@@ -669,9 +699,108 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+            const bool valid_chH = index_chH != 0;
+            if(valid_chH){
+                get(pfCand_chHad_valid) = valid_chH;
+                get(pfCand_chHad_rel_pt) = GetValueNorm(pfCands.at(index_chH).polarP4().pt() / tau.p4().pt(),
+                    is_inner ? 0.2564f : 0.0194f, is_inner ? 0.8607f : 0.1865f);
+                get(pfCand_chHad_deta) = GetValueLinear(pfCands.at(index_chH).polarP4().eta() - tau.p4().eta(),
+                    is_inner ? -0.1f : -0.5f, is_inner ? 0.1f : 0.5f, false);
+                get(pfCand_chHad_dphi) = GetValueLinear(dPhi(tau.p4(),pfCands.at(index_chH).polarP4()),
+                    is_inner ? -0.1f : -0.5f, is_inner ? 0.1f : 0.5f, false);
+                // get(pfCand_chHad_leadChargedHadrCand) = GetValue(dynamic_cast<const pat::PackedCandidate*>(tau.leadChargedHadrCand().get()));
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: here should check if a pointer to the current candidate equals to the pointer to leadChargedHadrCand. Similar to what is done here: https://github.com/kandrosov/TauML/blob/v2/Production/src/TauJet.cc#L145-L149

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 17 above)
Comment (MRD2F) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: what should be the condition? the same to check presence first, something like: 
`cell_map.count(CellObjectType::PfCand_chargedHadron)`

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 16 above)
Comment (MRD2F) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: you mean that should use at()?

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 17 above)
Comment (kandrosov) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: yes

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc (hunk 16 above)
Comment (kandrosov) [reply RecoTauTag/RecoTau/plugins/DeepTauId.cc]: yes

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 19]
```diff
@@ -669,9 +699,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+
+            const bool valid_chH = cell_map.count(CellObjectType::PfCand_chargedHadron);
+            const bool valid_nH = cell_map.count(CellObjectType::PfCand_neutralHadron);
+
+            if(valid_chH)
+                index_chH = cell_map.at(CellObjectType::PfCand_chargedHadron);
+            if(cell_map.count(CellObjectType::PfCand_neutralHadron))
+                index_nH = cell_map.at(CellObjectType::PfCand_neutralHadron);
+
+            if(valid_chH && valid_nH){
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: it is enough that one of them is valid, so `||`, not `&&`

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 20]
```diff
@@ -669,9 +699,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+                get(rho) = GetValueNorm(rho, 21.49f, 9.713f);
+                get(tau_pt) =  GetValueLinear(tau.polarP4().pt(), 20.f, 1000.f, true);
+                get(tau_eta) = GetValueLinear(tau.polarP4().eta(), -2.3f, 2.3f, false);
+                get(tau_inside_ecal_crack) = GetValue(isInEcalCrack(tau.p4().eta()));
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: use polarP4

File: RecoTauTag/RecoTau/plugins/DeepTauId.cc [hunk 21]
```diff
@@ -669,9 +699,104 @@ class DeepTauId : public deep_tau::DeepTauBase {
...
+            }
+            if(valid_chH){
+                get(pfCand_chHad_valid) = valid_chH;
+                get(pfCand_chHad_rel_pt) = GetValueNorm(pfCands.at(index_chH).polarP4().pt() / tau.p4().pt(),
+                    is_inner ? 0.2564f : 0.0194f, is_inner ? 0.8607f : 0.1865f);
+                get(pfCand_chHad_deta) = GetValueLinear(pfCands.at(index_chH).polarP4().eta() - tau.p4().eta(),
+                    is_inner ? -0.1f : -0.5f, is_inner ? 0.1f : 0.5f, false);
+                get(pfCand_chHad_dphi) = GetValueLinear(dPhi(tau.p4(),pfCands.at(index_chH).polarP4()),
```
Comment (kandrosov) [inline RecoTauTag/RecoTau/plugins/DeepTauId.cc]: use polarP4 in all 3 cases above.

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Review by kandrosov — APPROVED** [review]

**Pull request merged**